irohad/ametsuchi/impl/block_storage_nudb.cpp

```diff
@@ -0,0 +1,217 @@
+    }
+
+    bool BlockStorage::Impl::init(std::unique_ptr<nudb::store> db,
+                                  const std::string &path) {
+      db_ = std::move(db);
+
+      nudb::error_code ec;
+      total_blocks_ = count_blocks(ec);  // replaces check_consistency
+      // TODO(warchant): validate blocks during initialization
+
+      if (ec != nudb::error::key_not_found) {
+        log_->critical("can not read database to count blocks: {}",
+                       ec.message());
+        return false;
+      }
+
+      path_ = path;
+
+      return true;
+    }
+
+    bool BlockStorage::Impl::add(Identifier id,
+                                 const std::vector<uint8_t> &blob) {
+      nudb::error_code ec;
+      auto key = serialize_uint32(id);
+      db_->insert(key.data(), blob.data(), blob.size(), ec);
+      if (ec) {
+        log_->error("BlockStorage::add(): {}", ec.message());
+        return false;
+      }
+      ++total_blocks_;
+      return true;
+    }
+
+    boost::optional<std::vector<uint8_t>> BlockStorage::Impl::get(
+        Identifier id) const {
+      nudb::error_code ec;
+      boost::optional<std::vector<uint8_t>> ret;
+      auto key = serialize_uint32(id);
+      db_->fetch(key.data(),
+                 [&ret](const void *p, size_t size) {
+                   if (size == 0) {
+                     ret = boost::none;
+                   } else {
+                     const auto *c = static_cast<const char *>(p);
+                     ret = std::vector<uint8_t>{c, c + size};
+                   }
+                 },
+                 ec);
+      if (ec) {
+        log_->error("BlockStorage::get(): {}", ec.message());
+        return boost::none;
+      }
+
+      return ret;
+    }
+
+    bool BlockStorage::Impl::drop_db() {
+      nudb::error_code ec1, ec2, ec3;
+
+      nudb::erase_file(db_->dat_path(), ec1);
+      nudb::erase_file(db_->log_path(), ec2);
+      nudb::erase_file(db_->key_path(), ec3);
+
+      return !(ec1 || ec2 || ec3);
+    }
+
+    uint32_t BlockStorage::Impl::count_blocks(nudb::error_code &ec) {
+      BlockStorage::Identifier current = 0;
+
+      bool found_last = false;
+
+      do {
+        auto key = serialize_uint32(current);
+        db_->fetch(key.data(),
+                   [&found_last, &current](const void *value, size_t size) {
+                     // if we read 0 bytes, then there is no such key
+                     if (size == 0u) {
+                       found_last = true;
+                       return;
+                     }
+
+                     // go to the next key
+                     current++;
+                   },
+                   ec);
+        if (ec == nudb::error::key_not_found) {
+          return current;
+        } else if (ec) {
+          // some other error occurred
+          return 0;
+        }
+      } while (!found_last);
+
+      return current;
+    }
+
+    std::array<uint8_t, sizeof(uint32_t)> BlockStorage::Impl::serialize_uint32(
```

Comment:
Looks like it could be moved to utils.

Reply:
I agree, but maybe not now. We need separate class or set of utils, which can do serialization and deserialization. For now it is better to leave it here.